### DISCUSSION TITLE
is there any effect in zustand?

### DISCUSSION THREAD

**Discussion by leamandeep** [pmndrs/zustand#2246] [Q&A]
When JWT expires. I want to my user to auto logout. If there is something like side effect. It would be great. I don't know how to achieve side effect in zustand?

Below is my code.

```js
  setAuth: () => {
                const token = get().accessToken
                const decoded = jwtDecode(token)
                let isAuth = false
                const expDate = typeof decoded.exp !== "undefined" ? new Date(decoded.exp * 1000) : null;
                const iatDate = typeof decoded.iat !== "undefined" ? new Date(decoded.iat * 1000) : null;
                if (expDate && iatDate) {
                    if (iatDate < expDate) {
                        isAuth = true
                    } else if (iatDate > expDate) {
                        isAuth = false
                    } else {
                        isAuth = false
                    }
                } else {
                    isAuth = false
                }
                console.log(isAuth);
                set({ isAuth })
            }
        })
```

**Comment by dai-shi** [comment]
I'm not sure how you get notified when JWT expires, but you can basically update state outside React with `useFooStore.setState({ isAuth })`.

**Reply by leamandeep** [reply]
I have decoded jwt and extracted iat and exp date. Now i'm comparing both. But i don't know how to validate everytime when user is inside my app?
```js
                const token = get().accessToken
                const decoded = jwtDecode(token)
                let isAuth = false
                const expDate = typeof decoded.exp !== "undefined" ? new Date(decoded.exp * 1000) : null;
                const iatDate = typeof decoded.iat !== "undefined" ? new Date(decoded.iat * 1000) : null;
                if (expDate && iatDate) {
                    if (iatDate < expDate) {
                        isAuth = true
                    } else if (iatDate > expDate) {
                        isAuth = false
                    } else {
                        isAuth = false
                    }
                } else {
                    isAuth = false
                }
                console.log(isAuth);
                set({ isAuth })
```

**Reply by leamandeep** [reply]
I was looking for something like recoil. https://recoiljs.org/docs/guides/atom-effects

**Reply by dai-shi** [reply]
> when user is inside my app?

what does that mean? maybe you simply want to use useEffect.

**Comment by leamandeep** [comment]
> > when user is inside my app?
> 
> what does that mean? maybe you simply want to use useEffect.

I mean when user is on my website. At that time, if my jwt expires then my user will automatically logout. That's all
Yeah i was thinking to use useEffect

**Reply by dai-shi** [reply]
My question was whether you can only know it in React, or there's a way to know it outside React. Sorry for the confusion.

If `useEffect` works, it's probably a valid solution.
